runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5114,16 +5115,30 @@ class INTERPRETER_CLASS
 			}
 		}
 
+		bp = buildSpecialStackFrame(REGISTER_ARGS, J9SF_FRAME_TYPE_JNI_NATIVE_METHOD, jitStackFrameFlags(REGISTER_ARGS, 0), true);
+		*--_sp = (UDATA)_sendMethod;
+		_arg0EA = bp + 8;
 		updateVMStruct(REGISTER_ARGS);
 		VM_VMAccess::inlineExitVMToJNI(_currentThread);
+		VM_VMHelpers::beforeJNICall(_currentThread);
 #if FFI_NATIVE_RAW_API
 		ffi_ptrarray_to_raw(cif, values, values_raw);
 		ffi_raw_call(cif, FFI_FN(function), returnStorage, values_raw);
 #else /* FFI_NATIVE_RAW_API */
 		ffi_call(cif, FFI_FN(function), returnStorage, values);
 #endif /* FFI_NATIVE_RAW_API */
+		VM_VMHelpers::afterJNICall(_currentThread);
 		VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
 		VMStructHasBeenUpdated(REGISTER_ARGS);
+		if (VM_VMHelpers::exceptionPending(_currentThread)) {
```

Comment:
Is this possible (again, not calling JNI so who would set an exception pending?)?

Reply:
What happens if you downcall here and the C code upcalls back to java which throws an exception? Is the xception cleared before returning to the C code?

Reply:
> Is this possible? What happens if you downcall here and the C code upcalls back to java which throws an exception? Is the xception cleared before returning to the C code?

An exception (e.g. `OutOfMemory` when allocating an object) can be captured in the dispatcher `native2InterpJavaUpcallImpl` at [native2InterpJavaUpcallImpl](https://github.com/eclipse-openj9/openj9/blob/45a5c60888b51301325483de0e3f9f252f60b66b/runtime/vm/UpcallVMHelpers.cpp) in https://github.com/eclipse-openj9/openj9/pull/15309 which will be passed back here if the native function allows to do so (e.g. the native function return `0` for nothing in such case).

Reply:
I suspect this is invalid and should be explicitly tested. The whole point of this feature is to avoid interactions with JNI, so to me it doesn't make any sense for an exception to propagate through the call as there is no way to detect it without using JNI.

Reply:
I won't hold up the review for this, but you should explicitly test what occurs if an exception is thrown from one of these call-ins. My feeling is that the exception should simply be cleared before returning.

Reply:
The reason that the exception is checked here in the interpreter is that there is no way throw the exception directly in the native dispatcher at [native2InterpJavaUpcallImpl](https://github.com/eclipse-openj9/openj9/blob/45a5c60888b51301325483de0e3f9f252f60b66b/runtime/vm/UpcallVMHelpers.cpp) in https://github.com/eclipse-openj9/openj9/pull/15309, in which case it has to be carried back to the interpreter returning from the native function.
The only exception case is: when the exception occurs on a thread which is created locally in the dispatcher, the exception doesn't actually exist because the created thread will be cleaned up before returning to the interpreter.